Support calling __torch_function__ attribute access

mlazos opened the pull request:
Triggers `__torch_function__` tracing on attribute/method/property access matching the eager behavior for non-overridden attributes/methods/properties that are present on `torch.Tensor`.

Some caveats: 
1. for methods there doesn't seem to be a way to check if the original implementation of a method is overridden via monkey patching or not. For example:
```
class LocalSubclass(torch.Tensor):
    @classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        return super().__torch_function__(func, types, args, kwargs)

x = torch.ones(2, 2).as_subclass(LocalSubclass)


> x.sigmoid
<built-in method sigmoid of LocalSubclass object at 0x7f8d305bb5e0>
```
There isn't a way to verify that this built-in method is equivalent to the base `torch.Tensor` implementation as each instance will have a different built-in method object that can't be traced back to the original `torch.Tensor` impl. You can check that the class itself has the original implementation via 
```
> inspect.getattr_static(LocalSubclass, "sigmoid")
<method 'sigmoid' of 'torch._C.TensorBase' objects>
```
But we can't detect if the user dynamically patches an object with a built-in method called sigmoid which does something completely different. 

2. If a user overrides a method but calls the original implementation we will still graph break. This will require modifying `SuperVariable` (and any other way to get the original impl) to handle tensor subclasses. 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #111765
* __->__ #111737



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

mlazos committed:
Support calling __torch_function__ attribute access

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on test/dynamo/test_subclasses.py:
```diff
@@ -39,11 +66,15 @@ def __call__(self, gm: torch.fx.GraphModule, example_inputs):
+
+
 @contextlib.contextmanager
 def preserve_subclass_config():
```
Can't this just be `torch._dynamo.config.patch`

jansel commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -126,6 +126,30 @@ def subclass_type_var(self):
+            )
+
+        if tx.output.torch_function_enabled:
+            get_fn = SourcelessBuilder()(tx, getattr(torch.Tensor, name).__get__)
```
Check that `source` is not set?  Or are we sure no guards are needed?

mlazos replied:
I can create a source for this since I'm grabbing an attribute from `torch.Tensor`

jansel replied:
I think we need some sort of check that the user defined class doesn't overload `name`.  If the tensor subclass defines `name`, then `getattr(torch.Tensor, name)` will be wrong.

mlazos replied:
I checked if a property/method/attr is overridden on the class and trigger a graph break in this case. I added guards if the source is valid to ensure that the impl of a method/property/attr on the class matches the `torch.Tensor` impl as well. Handling Ed's case will be a bit trickier with `super()` handling for subclasses so I'm going to fallback for now.

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -45,6 +46,24 @@ def is_torch_function_user_object(obj):
     return hasattr(obj, "__torch_function__")
 
 
+def _is_attr_overidden(tx, var, name):
```
@jansel let me know what you think of this as a check

jansel commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -45,6 +46,24 @@ def is_torch_function_user_object(obj):
+        pass
+
+    try:
+        var.dynamic_getattr(tx, name)
```
Do we need to guard on this?

mlazos replied:
I removed the dynamic_getattr piece because it was giving some false positives. I added guards to ensure the an attribute/method/property is not overridden though if we make it past this check.

ezyang commented:
needs more pr desc?

ezyang commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1010,3 +1010,12 @@ def call_function(
+        return type(self.value)
+
+    def as_python_constant(self):
+        return self.value
```
UserDefinedObjectVariable does not report that it `is_python_constant`. Why should TensorSubclassVariable be any different?

mlazos replied:
This is more analagous to UserDefinedClassVariable, which is considered a constant

ezyang replied:
@jansel I suppose has the final word here, but this sounds like UserDefinedClassVariable is wrong to me. In general, Dynamo uses `is_python_constant` to test if something is a constant and thus Dynamo can directly manipulate it (e.g., putting these python constants into the graph, constant folding by directly working with the underlying Python value). An arbitrary defined user object never has this constant-ness; a user could have attached a metaclass to the class, there might be weird class variable mutation going on, etc.

jansel replied:
I tend to agree with that logic...
1) Are we guarding on ID to ensure it actually is a constant
2) Are we checking for metaclasses that could introduce side effects while compiling?
3) What code needs this is_python_constant stuff to work?

mlazos replied:
I'll remove for now, at least on my local run of the subclass tests it's passing.

mlazos committed:
Update on "Support calling __torch_function__ attribute access"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

mlazos committed:
Update on "Support calling __torch_function__ attribute access"

Triggers `__torch_function__` tracing on attribute/method/property access matching the eager behavior for non-overridden attributes/methods/properties that are present on `torch.Tensor`.

Some caveats: 
1. for methods there doesn't seem to be a way to check if the original implementation of a method is overridden via monkey patching or not. For example:
```
class LocalSubclass(torch.Tensor):
    classmethod
    def __torch_function__(cls, func, types, args=(), kwargs=None):
        if kwargs is None:
            kwargs = {}
        return super().__torch_function__(func, types, args, kwargs)

x = torch.ones(2, 2).as_subclass(LocalSubclass)


> x.sigmoid
<built-in method sigmoid of LocalSubclass object at 0x7f8d305bb5e0>
```
There isn't a way to verify that this built-in method is equivalent to the base `torch.Tensor` implementation as each instance will have a different built-in method object that can't be traced back to the original `torch.Tensor` impl. You can check that the class itself has the original implementation via 
```
> inspect.getattr_static(LocalSubclass, "sigmoid")
<method 'sigmoid' of 'torch._C.TensorBase' objects>
```
But we can't detect if the user dynamically patches an object with a built-in method called sigmoid which does something completely different. 

2. If a user overrides a method but calls the original implementation we will still graph break. This will require modifying `SuperVariable` (and any other way to get the original impl) to handle tensor subclasses. 




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 aakhundov kadeng

[ghstack-poisoned]

jansel commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -45,6 +46,19 @@ def is_torch_function_user_object(obj):
+        attr_val = inspect.getattr_static(var.python_type(), name)
+        overridden |= attr_val != getattr(torch.Tensor, name)
+    except AttributeError:
+        pass
```
When do we hit this case?   Is returning false correct?

mlazos replied:
Basically if a class overrides a property or method, `attr_val != getattr(torch.Tensor, name)` will be True, but if the class has the standard `torch.Tensor` implementation this will be False. I added test cases [here](https://github.com/pytorch/pytorch/pull/111737/files#:~:text=def%20test_user_overidden_attr_unsupported(self)%3A) to verify that this works.

ezyang replied:
Instead of catching AttributeError, you could do getattr(torch.Tensor, name, None)

ezyang commented on torch/_dynamo/variables/torch_function.py:
```diff
@@ -160,11 +213,24 @@ def call_method(
             # we will graph break in other cases this will need a bigger overhaul of extracting methods/comparing them for equality
-            func_var = SourcelessBuilder()(tx, getattr(torch.Tensor, name))
+            # We've established with the above check that the method is not overridden, so we guard that the method is the same
+            # as the impl defined on tensor and retrieve it
```
Hmm... this feels like over-guarding to me. This seems to imply that we will generate a guard for EVERY function access on the subclass to guard against the user post-facto monkeypatching their class to start overriding a method. It would be much better to do one of the version dict things we were doing for nn.Module to just test that, yes, the user hasn't modified their user defined class in the meantime.

ezyang reviewed (approved):
There are some parts that are dorky, especially the overguarding, but ok

mlazos commented:
@pytorchbot merge

mlazos commented:
@pytorchbot merge
